Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -141,17 +130,101 @@
+    if (!m_datagramConnection) {
+        ASSERT_NOT_REACHED();
+        return completionHandler(nullptr);
+    }
+
+    nw_connection_set_state_changed_handler(m_datagramConnection.get(), makeBlockPtr([
+        refThis = RefPtr { this },
+        completionHandler = WTFMove(completionHandler)
+    ] (nw_connection_state_t state, nw_error_t error) mutable {
+        if (!refThis)
+            return completionHandler(nullptr);
+        if (error)
+            return completionHandler(nullptr); // FIXME: Pipe the failure to JS
+        switch (state) {
+        case nw_connection_state_waiting:
+        case nw_connection_state_preparing:
+            return; // We will get another callback with another state change.
+        case nw_connection_state_ready:
+            return completionHandler(refThis.get());
+        case nw_connection_state_invalid:
+        case nw_connection_state_failed:
+        case nw_connection_state_cancelled:
+            return completionHandler(nullptr); // FIXME: Pipe the failure to JS
+        }
+        RELEASE_ASSERT_NOT_REACHED();
+    }).get());
+    nw_connection_set_queue(m_datagramConnection.get(), dispatch_get_main_queue());
+    nw_connection_start(m_datagramConnection.get());
+    receiveDatagramLoop();
+}
+
 void NetworkTransportSession::sendDatagram(std::span<const uint8_t> data, CompletionHandler<void()>&& completionHandler)
 {
-    nw_connection_group_send_message(m_connectionGroup.get(), makeDispatchData(Vector(data)).get(), m_endpoint.get(), NW_CONNECTION_DEFAULT_MESSAGE_CONTEXT, makeBlockPtr([completionHandler = WTFMove(completionHandler)] (nw_error_t error) mutable {
+    ASSERT(m_datagramConnection);
+    nw_connection_send(m_datagramConnection.get(), makeDispatchData(Vector(data)).get(), NW_CONNECTION_DEFAULT_MESSAGE_CONTEXT, true, makeBlockPtr([completionHandler = WTFMove(completionHandler)] (nw_error_t error) mutable {
+        ASSERT(!error);
         // FIXME: Pass any error through to JS.
         completionHandler();
     }).get());
 }
 
-void NetworkTransportSession::createBidirectionalStream(CompletionHandler<void(std::optional<WebTransportStreamIdentifier>)>&& completionHandler)
+void NetworkTransportSession::setupConnectionHandler()
 {
-    RetainPtr connection = adoptNS(nw_connection_group_extract_connection(m_connectionGroup.get(), m_endpoint.get(), nullptr));
+    nw_connection_group_set_new_connection_handler(m_connectionGroup.get(), makeBlockPtr([weakThis = WeakPtr { *this }] (nw_connection_t inboundConnection) mutable {
+        ASSERT_UNUSED(weakThis, weakThis);
+        // FIXME: Implement incoming connection handler
+    }).get());
+}
+
+void NetworkTransportSession::createStream(NetworkTransportStreamType streamType, CompletionHandler<void(std::optional<WebTransportStreamIdentifier>)>&& completionHandler)
+{
+    ASSERT(streamType != NetworkTransportStreamType::IncomingUnidirectional);
+#if HAVE(WEB_TRANSPORT)
```

Comment:
Could we have the entirety of this function in HAVE(WEB_TRANSPORT)?

#if HAVE(WEB_TRANSPORT)
...
#else
completionHandler(std::nullopt);
#endif

Same in NetworkTransportSession::setupDatagramConnection.  We might want to have more code inside WEB_TRANSPORT checks, and just stubs outside.  That may also require running "run-webkit-tests imported/w3c/web-platform-tests/webtransport --reset --no-build" on a system without WEB_TRANSPORT to keep the test bots green, and we may need internal test expectations for these tests.